### PR TITLE
Side-metadata supported object remembering barrier

### PR DESCRIPTION
**Pull request by wenyuzhao**
Related to https://github.com/mmtk/mmtk-core/pull/183

* Metadata memory range:
    * **Metadata start**: `HEAP_END`
    * **Metadata pages**: `MAX_CHUNKS * SelectedConstraints::METADATA_PAGES_PER_CHUNK`
* Lazily initialize metadata memory for each chunk, during its SFT initialization.

[**[Performance Comparison Results]**](https://squirrel.anu.edu.au/plotty/wenyuz/rust-g1/#0|shrew-2020-12-10-Thu-214900&benchmark^build^hfac^invocation^iteration&time^time.gc^time.mu&|10&iteration^1^4&benchmark^2^luindex|20&1^invocation|30&1&benchmark&build;jdk-mmtk-barrier-hashset|70&build^jdk|41&Histogram%20(with%20CI)^build^benchmark&)

# TODO

* [x] ~Performance comparison~
* [x] Clear metadata as a whole
* [x] Finalize side-metadata design with @javadamiri

### CODE REVIEW

File: src/policy/space.rs
Comment (wenyuzhao) [inline src/policy/space.rs:585]: This code eagerly mmaps all the metadata memory for the whole 64-bit contiguous space, which will most likely panic due to OOM.

File: src/util/heap/monotonepageresource.rs
Comment (wenyuzhao) [inline src/util/heap/monotonepageresource.rs:177]: Map metadata memory only when the heap chunk is not mmaped before. See fragmented_mapper.rs

File: src/util/heap/layout/fragmented_mapper.rs
Comment (wenyuzhao) [inline src/util/heap/layout/fragmented_mapper.rs:110]: Call the callback function only when the chunk is not mapped before.

Comment (javadamiri) [reply src/util/heap/layout/fragmented_mapper.rs:110]: I suggest adding two arguments to `Mmapper::ensure_mapped` e.g. change it to `Mmapper::ensure_mapped(start, pages, global_metadata_per_chunk: usize, local_metadata_per_chunk: usize)`, and probably do the same for `dzmmap()`.

Comment (javadamiri) [reply src/util/heap/layout/fragmented_mapper.rs:110]: This will be slightly different if we switch from per-chunk side metadata, but is conceptually the same

**Comment by qinsoon** [comment]
@wenyuzhao Can you fix the check?

**Comment by wenyuzhao** [comment]
> @wenyuzhao Can you fix the failed check?

Yes I'll do that asap. I think a more urgent thing is that we need to figure out a better way to replace my metadata fix in this PR, as I pointed out in the comments.

**Comment by qinsoon** [comment]
> > @wenyuzhao Can you fix the failed check?
> 
> Yes I'll do that asap. I think a more urgent thing is that we need to figure out a better way to replace my metadata fix in this PR, as I pointed out in the comments.

I agree. I was asking just to avoid a deadlock issue - I am waiting for your fix for the test while you are waiting for my review.

**Comment by qinsoon** [comment]
openjdk-binding-test
OPENJDK_BINDING_REF=gencopy-object-remembering-barrier-fastpath

File: src/plan/gencopy/mod.rs
Comment (qinsoon) [inline src/plan/gencopy/mod.rs:9]: Are those 3 constants only used for you to run our experiments? If my understanding is correct, they change the behavior of gen copy, and they should not be set to `true` for a normal user. If that's the case, I would suggest removing them from the code. It is more about the methodology you used for the experiments.

File: src/plan/barriers.rs
Comment (qinsoon) [inline src/plan/barriers.rs:44]: Though you pass in a constant for `SideMetadataSpec`, it probably is not treated as a constant when you use the field `meta`. Does it matter for performance?

File: src/plan/gencopy/gc_work.rs
Comment (qinsoon) [inline src/plan/gencopy/gc_work.rs:213]: Was this code unused?

File: src/plan/gencopy/global.rs
Comment (qinsoon) [inline src/plan/gencopy/global.rs:176]: Can you explain this change? Previously the plan needed 2 bits global side metadata if we don't have GC bytes, and needed 0 if we have GC bytes. Now the plan needed 1 bit and assumed we always have GC bytes.  Is the 1 bit for `LOGGING_META`? What's the reason we need to assume we always have GC bytes for GenCopy?

File: src/scheduler/gc_work.rs
Comment (qinsoon) [inline src/scheduler/gc_work.rs:456]: This `meta` also will not be treated as a constant.

File: src/scheduler/gc_work.rs
Comment (qinsoon) [inline src/scheduler/gc_work.rs:481]: This will do the work with the current thread and the current work packet. What is the reason for this change? Does  this affect the correctness of work packet statistics? Basically the `ScanObject` work is counted in a `ProcessModBuf` work.

File: src/util/heap/layout/byte_map_mmapper.rs
Comment (qinsoon) [inline src/util/heap/layout/byte_map_mmapper.rs:209]: This does not make sense? You can rename it to `CHUNK_SIZE`.

File: src/util/heap/freelistpageresource.rs
Comment (qinsoon) [inline src/util/heap/freelistpageresource.rs:120]: Why is the callback not invoked for `FreelistPageResource`?

File: src/policy/space.rs
Comment (qinsoon) [reply src/policy/space.rs:585]: In the `init()` for `LockFreeImmortalSpace`, we also have code like this.

File: src/util/heap/layout/vm_layout_constants.rs
Comment (qinsoon) [inline src/util/heap/layout/vm_layout_constants.rs:90]: What is the reason for changing this?

File: src/util/heap/layout/fragmented_mapper.rs
Comment (qinsoon) [reply src/util/heap/layout/fragmented_mapper.rs:110]: Why is this callback only used in `FragmentedMapper` (64-bit only)? You removed the side metadata mmap code from `Space.init()` which is used for both 32bit and 64bit. Does this mean we are not mmapping data for contiguous spaces in 32 bits?

File: src/util/heap/layout/fragmented_mapper.rs
Comment (qinsoon) [inline src/util/heap/layout/fragmented_mapper.rs:356]: This is related to my question: why we need to change the mmap chunk size. But generally, it is not good to just comment out failed assertions. It usually reveals some problems if we have failed assertions - it could be the test itself that was right.

File: src/util/heap/layout/vm_layout_constants.rs
Comment (wenyuzhao) [reply src/util/heap/layout/vm_layout_constants.rs:90]: `LOG_MMAP_CHUNK_BYTES = 20` means you have to do 4 mmap calls to map a 4mb chunk. And Mmappers (e.g. FragmentedMapper) have a table to track the mapping state for each MMAP_CHUNK. So it can map the 1mb MMAP_CHUNK only if it is not mapped before.

By setting the size of a MMAP_CHUNK equals to a 4MB CHUNK, we can directly use the mapping state table to check whether we need to map metadata memory as well.

File: src/util/heap/freelistpageresource.rs
Comment (wenyuzhao) [reply src/util/heap/freelistpageresource.rs:120]: This is just my temporary metadata quick fix. I only made it work with 64bit contiguous space, so this code will never be executed when running gencopy with openjdk. Anyway, we need to fix this after figuring out a better approach for the fix.

File: src/util/heap/layout/vm_layout_constants.rs
Comment (qinsoon) [reply src/util/heap/layout/vm_layout_constants.rs:90]: @steveblackburn Is there any reason that a mmap chunk is 1MB and a chunk is 4MB? Is it okay if we change the mmap chunk size to 4MB as well?

File: src/plan/gencopy/global.rs
Comment (wenyuzhao) [reply src/plan/gencopy/global.rs:176]: If openjdk turns off gc-byte support, then the gc byte has to be moved to side-metadata, which means the barrier also needs to be updated to adapt to the new layout... So I just add an assertion here to check this can never happen.

File: src/util/heap/freelistpageresource.rs
Comment (qinsoon) [reply src/util/heap/freelistpageresource.rs:120]: Alright. You requested reviews so I assumed it was ready for reviewing.

File: src/plan/gencopy/global.rs
Comment (qinsoon) [reply src/plan/gencopy/global.rs:176]: If it is for openjdk, you probably need to keep the assertions at the openjdk side.

File: src/plan/gencopy/mod.rs
Comment (wenyuzhao) [reply src/plan/gencopy/mod.rs:9]: I suggest we keep these flags. These flags provide a way to run GenCopy without any barriers, or with a completely different barrier. With these flags, one can directly clone mmtk-core and openjdk binding, and start doing barrier overhead measurements without further changes to the GenCopy GC, which is quite convenient.  We also need this feature later for barrier fast path debugging.

File: src/util/heap/layout/fragmented_mapper.rs
Comment (wenyuzhao) [reply src/util/heap/layout/fragmented_mapper.rs:110]: One design problem of this fix is, `Space` is the caller of this mmapper code. When a space is notified by a mmaper (using `notify_mmap`), it needs to map both space-specific and global metadata. But spaces shouldn't access/manage global metadata by design.

Comment (wenyuzhao) [reply src/util/heap/layout/fragmented_mapper.rs:110]: Anyway, we still need to refactor this mmap code at some point in the future.  When a mmtk heap chunk is mapped by some third-party other than mmtk, we should not do metadata memory mapping.

File: src/plan/barriers.rs
Comment (wenyuzhao) [reply src/plan/barriers.rs:44]: I guess it is possible to treat them as constants if we pass them as a set of associated constants wrapped in a generic type parameter. I'll fix this (and the other one).

File: src/plan/gencopy/gc_work.rs
Comment (wenyuzhao) [reply src/plan/gencopy/gc_work.rs:213]: The code is unused at the very beginning (when GenCopy is merged for the first time). GenCopy uses `scheduler::gc_work::ProcessModBuf` instead.

File: src/plan/gencopy/mod.rs
Comment (qinsoon) [reply src/plan/gencopy/mod.rs:9]: Okay. If we have to keep them, I suggest adding some more comments:
1. For where the constants are defined, explain how to config those constants to measure overhead for different components
2. For where those constants are used, be very clear that the constant is only used for special methodology for measuring overhead of different components. The reason why I suggest removing them is those conditional checks are quite disruptive when I tried to read the code, as some essential code will be skipped based on the constant - which is confusing.

File: src/scheduler/gc_work.rs
Comment (wenyuzhao) [reply src/scheduler/gc_work.rs:481]: This work-packet only produces one (instead of `N`) ScanObjects packets. So I think it makes no sense to push it to the local work-packet cache and pull it pack before executing it.

Yes it will affect the statistics. Thanks for pointing this out. I'll fix this.

File: src/plan/barriers.rs
Comment (qinsoon) [reply src/plan/barriers.rs:44]: I was just pointing out this could potentially be an issue. If in your measurement, this is not an issue then we dont need to do anything for it.

File: src/plan/barriers.rs
Comment (wenyuzhao) [reply src/plan/barriers.rs]: `nursery` is not used anymore. I've removed it.

File: src/plan/barriers.rs
Comment (wenyuzhao) [reply src/plan/barriers.rs]: Done. I removed all the barrier counter code.

File: src/plan/barriers.rs
Comment (wenyuzhao) [reply src/plan/barriers.rs]: Removed

File: src/scheduler/gc_work.rs
Comment (wenyuzhao) [reply src/scheduler/gc_work.rs:481]: I double checked the code. By calling `GCWork::do_work(&mut ScanObjects...)` directly, the `ScanObjects` packet is not accounted when collecting statistics.

File: src/util/heap/layout/byte_map_mmapper.rs
Comment (wenyuzhao) [reply src/util/heap/layout/byte_map_mmapper.rs]: Done

File: src/policy/space.rs
Comment (wenyuzhao) [reply src/policy/space.rs:585]: I'm not quite sure how to deal with `LockFreeImmortalSpace`. It maps all the heap memory ahead of time. But mapping metadata memory ahead of time as well will most likely fail. Probably we shouldn't have any metadata when using  `LockFreeImmortalSpace`.

File: src/util/heap/layout/vm_layout_constants.rs
Comment (steveblackburn) [reply src/util/heap/layout/vm_layout_constants.rs:90]: I don't know the history behind that particular decision.

However, the danger is that this will cause us to significantly overstate our physical memory use.

File: src/util/heap/layout/fragmented_mapper.rs
Comment (wenyuzhao) [reply src/util/heap/layout/fragmented_mapper.rs]: > I suggest adding two arguments to `Mmapper::ensure_mapped` e.g. change it to `Mmapper::ensure_mapped(start, pages, global_metadata_per_chunk: usize, local_metadata_per_chunk: usize)`, and probably do the same for `dzmmap()`.

Fixed

File: src/util/heap/freelistpageresource.rs
Comment (wenyuzhao) [reply src/util/heap/freelistpageresource.rs]: Fixed. Now support FreelistPageResource as well.

File: src/policy/space.rs
Comment (wenyuzhao) [reply src/policy/space.rs:585]: We don't need to do anything to `LockFreeImmortalSpace`. It maps metadata only for the heap that the VM actually uses.

File: src/plan/gencopy/mod.rs
Comment (wenyuzhao) [reply src/plan/gencopy/mod.rs]: Done.

File: src/plan/gencopy/global.rs
Comment (wenyuzhao) [reply src/plan/gencopy/global.rs]: Done. The check is now moved to openjdk binding

File: src/util/heap/layout/fragmented_mapper.rs
Comment (wenyuzhao) [reply src/util/heap/layout/fragmented_mapper.rs]: This test is fixed.

**Review by qinsoon — APPROVED** [review]

**Comment by qinsoon** [comment]
Is this PR ready to merge? @wenyuzhao

**Comment by wenyuzhao** [comment]
> Is this PR ready to merge? @wenyuzhao

Yes it's ready to merge, after the mmtk-openjdk PR is approved.

**Pull request merged**
